cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -150,6 +150,80 @@ void AddListCast(CastFunction* func) {
   DCHECK_OK(func->AddKernel(SrcType::type_id, std::move(kernel)));
 }
 
+struct CastStruct {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const CastOptions& options = CastState::Get(ctx);
+    const auto in_field_count =
+        checked_cast<const StructType&>(*batch[0].type()).num_fields();
+    const auto out_field_count =
+        checked_cast<const StructType&>(*out->type()).num_fields();
+
+    if (in_field_count != out_field_count) {
+      return Status::TypeError("struct field sizes do not match: ",
+                               batch[0].type()->ToString(), " ", out->type()->ToString());
+    }
+
+    for (int64_t i = 0; i < in_field_count; ++i) {
```

Comment:
MSVC is a bit stricter about comparing integers of the same width/sign so it might be easier to use `int` here to keep types consistent